harness/determined/pytorch/_pytorch_trial.py

```diff
@@ -535,16 +535,24 @@ def _compute_validation_metrics(self) -> workload.Response:
                 metrics = self._convert_metrics_to_numpy(metrics)
                 num_inputs = self.context.get_per_slot_batch_size() * len(self.validation_loader)
 
-        # TODO(yoavz): If any validation step end callbacks are defined,
-        # broadcast metrics across all shards here and execute the callback
-        # function on all shards.
-
-        if not self.is_chief:
-            return workload.Skipped()
+        if self.hvd_config.use and any(
+            map(
+                lambda c: util.is_overridden(c.on_validation_step_end, _callback.PyTorchCallback),
+                self.callbacks.values(),
+            )
+        ):
+            logging.debug(
+                "Broadcasting metrics to all worker processes to execute a "
+                "validation step end callback"
+            )
+            metrics = hvd.broadcast_object(metrics, root_rank=0)
```

Comment:
Is is possible that if users are using `evaluate_full_dataset` their metrics are not broadcastable. If this is the case, should we do something like:

```
try:
    broadcasted_metrics = hvd.broadcast_object(metrics, root_rank=0)
except Exception as e:
    if self._evaluate_full_dataset_defined():
         logging.warning("...")
         broadcasted_metrics = metrics if self.is_chief else {}
     else:
         raise e
finally:
    metrics = broadcasted_metrics
```

Reply:
Hmmm, I was thinking we already have some restrictions that metrics are JSON-serializable. Eventually we need to serialize these metrics to send from harness -> master and then to serialize to DB. 

Looking at the source of `broadcast_object`, it uses `cloudpickle` which can definitely serialize anything that is already JSON-serializable. So in other words, if it fails on broadcasting, seems pretty likely it would fail down our stack regardless.

Reply:
Let me investigate how these failures surface and get back to you